src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```
@@ -0,0 +1,199 @@
+
+        valid_name = to_identifier(database_name)
+
+        try:
+            db_result = self._sql_executor.execute_query(
+                "select current_database()"
+            ).fetchone()
+        except Exception as err:
+            return handle_unclassified_error(err, "Failed to select current database.")
+
+        try:
+            prev_db = db_result[0]
+        except:
+            prev_db = None
+
+        is_different_db = valid_name != prev_db
+        if is_different_db:
+            self._log.debug(f"Using database {valid_name}")
+            self._use_object(ObjectType.DATABASE, valid_name)
+
+        try:
+            yield
+        finally:
+            if is_different_db and prev_db is not None:
+                self._log.debug(f"Switching back to database:{prev_db}")
+                self._use_object(ObjectType.DATABASE, prev_db)
+
+    def execute_user_script(
+        self,
+        queries: str,
+        script_name: str,
+        role: str | None = None,
+        warehouse: str | None = None,
+        database: str | None = None,
+    ):
+        """
+        Runs the user-provided sql script.
+        @param queries: Queries to run in this script
+        @param script_name: Name of the file containing the script. Used to show logs to the user.
+        @param [Optional] role: Role to switch to while running this script. Current role will be used if no role is passed in.
+        @param [Optional] warehouse: Warehouse to use while running this script.
+        @param [Optional] database: Database to use while running this script.
+        """
+        with self._use_role_optional(role):
+            with self._use_warehouse_optional(warehouse):
+                with self._use_database_optional(database):
+                    try:
+                        self._sql_executor.execute_queries(queries)
+                    except ProgrammingError as err:
+                        if err.errno == NO_WAREHOUSE_SELECTED_IN_SESSION:
```

Is this a special case of only some queries or was there a decision made to let it fail here? It feels a bit counterintuitive to me to say that `warehouse` parameter is optional and then we expect the call to fail if no warehouse was selected

---

WH is optional yes, but we don't know if the SQL commands will require one. So they can fail due to the user not having provided a WH. IMHO that's a fair failure mode, the user is escaping to SQL, they should know whether a WH is needed or not. We shouldn't force them to create/activate one just in case they need it.